File: js/cmd/key-gen.js
import * as ed25519 from '@ucanto/principal/ed25519'
import * as rsa from '@ucanto/principal/rsa'
import * as dagJSON from '@ipld/dag-json'

/** @param {{ type?: 'rsa'|'ed25519' }} opts */
export const keyGen = async opts =>
  console.log(dagJSON.stringify(opts.type === 'rsa' ? await genRSA() : await genEd25519()))

const genEd25519 = async () => {
  const signer = await ed25519.generate()
  return { id: signer.did(), key: signer.encode() }
}

const genRSA = async () => {
  const signer = await rsa.generate({ extractable: true })
  const archive = signer.toArchive()
  return { id: signer.did(), key: archive.keys[archive.id] }
}

File: go/cmd/key-gen.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/ipld/go-ipld-prime"
	"github.com/ipld/go-ipld-prime/codec/dagjson"
	ed25519 "github.com/storacha/go-ucanto/principal/ed25519/signer"
	rsa "github.com/storacha/go-ucanto/principal/rsa/signer"
	"github.com/urfave/cli/v2"
)

type keyGenOut struct {
	id  string
	key []byte
}

func KeyGen(cCtx *cli.Context) error {
	var id string
	var key []byte
	if strings.ToLower(cCtx.String("type")) == "rsa" {
		signer, err := rsa.Generate()
		if err != nil {
			return err
		}
		id = signer.DID().String()
		key = signer.Encode()
	} else {
		signer, err := ed25519.Generate()
		if err != nil {
			return err
		}
		id = signer.DID().String()
		key = signer.Encode()
	}

	json, err := ipld.Marshal(dagjson.Encode, &keyGenOut{id, key}, nil)
	if err != nil {
		return fmt.Errorf("encoding response: %s", err)
	}

	fmt.Println(string(json))
	return nil
}
